File: go/leetcode/easy/linkedListCycle.go
// https://leetcode.com/problems/linked-list-cycle/

func hasCycle(head *ListNode) bool {
    if head == nil {
        return false
    }

    visited := make(map[*ListNode]bool)
    return hasCycleWithVisited(head.Next, visited)
}

func hasCycleWithVisited(head *ListNode, visited map[*ListNode]bool) bool {
    if head == nil {
        return false
    }

    if visited[head] {
        return true
    }
    visited[head] = true
    return hasCycleWithVisited(head.Next, visited)
}

File: python/leetcode/easy/linked_list_cycle.py
# https://leetcode.com/problems/linked-list-cycle/

class Solution:
    def __init__(self):
        self.visited = {}

    def hasCycle(self, head: ListNode) -> bool:
        if head is None:
            return False

        if head in self.visited:
            return True

        self.visited[head] = 1
        return self.hasCycle(head.next)
